controllers/effective_tls_policies_reconciler.go

```diff
@@ -0,0 +1,394 @@
+			l, ok := t.(*machinery.Listener)
+			return l, ok && lo.Contains(l.Gateway.Policies(), p)
+		})
+
+		// Policy is deleted
+		if policy.DeletionTimestamp != nil {
+			logger.V(1).Info("policy is marked for deletion, nothing to do", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+			continue
+		}
+
+		// Policy is not valid
+		isValid, _ := IsTLSPolicyValid(ctx, s, policy)
+		if !isValid {
+			logger.V(1).Info("deleting certs for invalid policy", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+			if err := t.deleteCertificatesForPolicy(ctx, topology, policy); err != nil {
+				logger.Error(err, "unable to delete certs for invalid policy", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+			}
+			continue
+		}
+
+		// Policy is valid
+		for _, l := range listeners {
+			// Need to use Gateway as listener hosts can be merged into a singular cert if using the same cert reference
+			expectedCertificates := expectedCertificatesForGateway(ctx, l.Gateway.Gateway, policy)
+
+			for _, cert := range expectedCertificates {
+				resource := t.client.Resource(CertManagerCertificatesResource).Namespace(cert.GetNamespace())
+
+				// Check is cert already in topology
+				objs := topology.Objects().Children(l)
+				obj, ok := lo.Find(objs, func(o machinery.Object) bool {
+					return o.GroupVersionKind().GroupKind() == CertManagerCertificateKind && o.GetNamespace() == cert.GetNamespace() && o.GetName() == cert.GetName()
+				})
+
+				// Create
+				if !ok {
+					expectedCerts = append(expectedCerts, cert)
+					if err := controllerutil.SetControllerReference(policy, cert, t.scheme); err != nil {
+						logger.Error(err, "failed to set owner reference on certificate", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+						continue
+					}
+
+					un, err := controller.Destruct(cert)
+					if err != nil {
+						logger.Error(err, "unable to destruct cert")
+						continue
+					}
+					_, err = resource.Create(ctx, un, metav1.CreateOptions{})
+					if err != nil && !apierrors.IsAlreadyExists(err) {
+						logger.Error(err, "unable to create certificate", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+					}
+
+					continue
+				}
+
+				// Update
+				tCert := obj.(*controller.RuntimeObject).Object.(*certmanv1.Certificate)
+				expectedCerts = append(expectedCerts, tCert)
+				if reflect.DeepEqual(tCert.Spec, cert.Spec) {
+					logger.V(1).Info("skipping update, cert specs are the same, nothing to do")
+					continue
+				}
+
+				tCert.Spec = cert.Spec
+				un, err := controller.Destruct(tCert)
+				if err != nil {
+					logger.Error(err, "unable to destruct cert")
+					continue
+				}
+				_, err = resource.Update(ctx, un, metav1.UpdateOptions{})
+				if err != nil {
+					logger.Error(err, "unable to update certificate", "policy", policy.Name)
+				}
+			}
+		}
+	}
+
+	// Clean up orphaned certs
```

Comment:
I believe this is because there is a indirect link between policies and the subresouce (certificates in this case) and so gives the expected state of the cluster. Though true, this probably can be done as a separate task, that I can look into if you feel strongly about this. 

A cert can be orphaned when a gateway listener is removed, or is the target ref of a tls policy is changed to another gateway. In both of these cases the cert(s) is created and will be orphaned since the policy still exists and still valid

Reply:
Indeed we need to cleanup certs that remained orphan due to, e.g., a listener being removed from a gateway _ceteris paribus_.

Let's keep this as-is now since it does the job. Maybe in the future we can look for this pattern across the reconcilers (not only this one) and try to improve them all by having "janitors" as separate tasks that can run in parallel to other tasks.

Reply:
I am happy with leaving this for now. It is some thing that can be done in the future.